### PR TITLE
Add coroutines await() to signals

### PR DESCRIPTION
**Pull request by CedNaru**
- Add a new godot.couroutine module.
- The Gradle plugin can now optionnally include it (and implicitly `kotlinx.coroutines`)
- Add a Godot specific scope for coroutines
- Add an `await()` functions to all signals that will suspend the coroutines until the signal is emitted.
- `await()` can return the signal's arguments, directly (when 1), as a tuple (when 2 and 3), and as list (4+)

I also changed the package of Signal and Callable (I needed them for the await implementation so I took the opportunity). 
So far everything has been in the same `godot.core` package except for those 2. Maybe it's an error and we should split everything into smaller packages but right now I wanted to focus on more consistency.
Note that the await() functions are in the `godot.coroutines` package as it's not considered the same library.

### CODE REVIEW

File: harness/tests/test/unit/test_coroutines.gd [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    test_script.start_coroutine_with_one_parameter()
+    assert_eq(test_script.step, 3, "Property should be 3 after coroutine started but waiting.")
+
+    test_script.signal_with_one_parameter.emit(4)
+    await get_tree().create_timer(1).timeout
+    assert_eq(test_script.step, 4, "Property should be 4 after coroutine ran.")
+
+    test_script.start_coroutine_with_many_parameters()
+    assert_eq(test_script.step, 5, "Property should be 3 after coroutine started but waiting.")
```
Comment (piiertho) [inline harness/tests/test/unit/test_coroutines.gd:24]: Error in message. Property should be 5

File: harness/tests/test/unit/test_coroutines.gd [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    test_script.signal_with_many_parameters.emit(6, 0.1, Vector2(0,0), "test")
+    await get_tree().create_timer(1).timeout
+    assert_eq(test_script.step, 6, "Property should be 4 after coroutine ran.")
```
Comment (piiertho) [inline harness/tests/test/unit/test_coroutines.gd:28]: Same here

File: kt/plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/GodotPlugin.kt [hunk 3]
```diff
@@ -18,24 +21,30 @@ abstract class GodotPlugin : Plugin<Project> {
...
+        afterEvaluate {
+            if (godotJvmExtension.enableGodotCoroutines.get()) {
+                dependencies.add(
+                    "implementation",
+                    dependencies.create("com.utopia-rise:${godotCoroutineLibraryArtifactName}:${GodotBuildProperties.assembledGodotKotlinJvmVersion}")
+                )
+            }
             registry.register(
                 PropertiesModelBuilder(
                     isFqNameRegistrationEnabled = extension.isFqNameRegistrationEnabled.get(),
```
Comment (chippmann) [inline kt/plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/GodotPlugin.kt:45]: Please leave the explicit target be. There were are some scoping issues before and i would not like to introduce such issues again

**Review by chippmann — CHANGES_REQUESTED** [review]
Did you check that none of the IDE checks are requiring the signals to be in the signal package?

File: kt/api-generator/src/main/kotlin/godot/codegen/extensions/TypedExtensions.kt [hunk 4]
```diff
@@ -61,10 +60,10 @@ fun TypedTrait.isTypedArray() = type?.startsWith(GodotTypes.typedArray) ?: false
 fun TypedTrait.getTypeClassName(): ClassTypeNameWrapper {
     val typeNameWrapper = when {
         type.isNullOrEmpty() -> ClassTypeNameWrapper(UNIT)
-        type == "Signal0" -> ClassTypeNameWrapper(ClassName(signalPackage, type!!))
+        type == "Signal0" -> ClassTypeNameWrapper(ClassName(godotCorePackage, type!!))
         type!!.startsWith("Signal") -> {
             this as EnrichedSignal
-            val className = ClassName(signalPackage, type)
```
Comment (piiertho) [inline kt/api-generator/src/main/kotlin/godot/codegen/extensions/TypedExtensions.kt:67]: Was ˋsignalPackage` variable deleted ?

File: kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+import com.squareup.kotlinpoet.TypeSpec
+import com.squareup.kotlinpoet.TypeVariableName
+import com.squareup.kotlinpoet.UNIT
+import godot.codegen.services.IAwaitGenerationService
+import godot.tools.common.constants.GODOT_OBJECT
+import godot.tools.common.constants.GodotKotlinJvmTypes.signal
+import godot.tools.common.constants.godotCorePackage
+import godot.tools.common.constants.godotCoroutinePackage
+
+private val cancellableContinuationClass = ClassName("kotlinx.coroutines", "CancellableContinuation")
```
Comment (piiertho) [inline kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt:21]: I would set `kotlinx.coroutines` as a constant

File: kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+                    }
+                    .generateBody(argCount, returnType)
+                    .returns(returnType)
+                    .build()
+            )
+        }
+
+        return awaitFile
+            .addAnnotation(
+                AnnotationSpec.builder(ClassName("kotlin", "Suppress"))
```
Comment (piiertho) [inline kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt:95]: You can use `Suppress` class instead of strings here.

File: kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+                    .addMember("\"PackageDirectoryMismatch\", \"unused\"")
+                    .build()
+            )
+            .indent("    ")
```
Comment (piiertho) [inline kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt:99]: Why is this needed ?

File: kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        val lambdaParameters = buildString {
+            for (i in 0 until argCount) {
+                if (i != 0) {
+                    append(", ")
+                }
+                append("p$i")
+            }
+        }
+
+        // Build what is inserted into the `resume()` method : `Unit`, `po`, `Pair(p0, P1)`, `Triple(p0, p1, p2)`, `listOf(p0, p1, p2, p3), etc..`
```
Comment (piiertho) [inline kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt:115]: Comment not anymore true as far as I understand.

File: kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+            .build()
+    }
+
+
+    private fun FunSpec.Builder.generateBody(argCount: Int, returnType: TypeName): FunSpec.Builder {
+        // Build `p0, p1, ..., px`
+        val lambdaParameters = buildString {
+            for (i in 0 until argCount) {
+                if (i != 0) {
+                    append(", ")
```
Comment (piiertho) [inline kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt:109]: You should replace space with `·`

File: kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+        // Build what is inserted into the `resume()` method : `Unit`, `po`, `Pair(p0, P1)`, `Triple(p0, p1, p2)`, `listOf(p0, p1, p2, p3), etc..`
+        val resumeParameters = when (argCount) {
+            0 -> "Unit"
+            1 -> lambdaParameters
+            else -> "SignalArguments$argCount($lambdaParameters)"
+        }
+
+        return this
+            .beginControlFlow("return %M", suspendCancellableCoroutine)
```
Comment (piiertho) [inline kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt:123]: Same here and below, spaces should be replaced

File: kt/godot-library/src/main/kotlin/godot/core/SignalProvider.kt [hunk 11]
```diff
@@ -1,22 +1,22 @@
-package godot.signals
```
Comment (piiertho) [inline kt/godot-library/src/main/kotlin/godot/core/SignalProvider.kt:1]: Why the whole file is rewritten ?

File: kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt (hunk 7 above)
Comment (CedNaru) [reply kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt:99]: Because Kotlin Poet was using indent of 2 spaces, and somehow was failing the formatting of the code, which led to it not compiling.

File: kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+                baseReceiver
+            }
+
+            val returnType: TypeName = when (argCount) {
+                0 -> UNIT
+                1 -> parameters[0]
+                else -> ClassName(godotCoroutinePackage, "SignalArguments$argCount").parameterizedBy(parameters)
+            }
+
+            // Create a tuple for the signal arguments
```
[the remark this answers is not in the document]
Comment (CedNaru) [reply kt/api-generator/src/main/kotlin/godot/codegen/services/impl/AwaitGenerationService.kt:50]: It's literally a tuple

File: kt/godot-coroutine-library/src/main/kotlin/godot/coroutines/GodotCoroutine.kt [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+object GodotCoroutine {
+    private val scope = CoroutineScope(Dispatchers.Default)
+
+    operator fun invoke(block: suspend CoroutineScope.() -> Unit) {
+        scope.launch(start = CoroutineStart.UNDISPATCHED, block = block)
+    }
+}
```
[the remark this answers is not in the document]
Comment (CedNaru) [reply kt/godot-coroutine-library/src/main/kotlin/godot/coroutines/GodotCoroutine.kt:14]: Updated according to what was said on Discord

File: kt/godot-coroutine-library/build.gradle.kts [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+tasks {
+    compileKotlin {
+        dependsOn(":godot-library:generateAPI")
+    }
+
+    // here so the sourcesJar task has an explicit dependency on the generateApi task. Needed since gradle 8
+    withType<Jar> {
+        dependsOn(":godot-library:generateAPI")
+    }
+}
```
[the remark this answers is not in the document]
Comment (CedNaru) [reply kt/godot-coroutine-library/build.gradle.kts:44]: I tried to remove it but it doesn't want to build without that.

I get the following error:

What went wrong:
A problem was found with the configuration of task ':kt:godot-coroutine-library:sourcesJar' (type 'Jar').
- Gradle detected a problem with the following location: 'D:\Godot\Module\kotlin\modules\kotlin_jvm\kt\godot-coroutine-library\src\main\kotlin'.

  Reason: Task ':kt:godot-coroutine-library:sourcesJar' uses this output of task ':kt:godot-library:generateAPI' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.

  Possible solutions:
    1. Declare task ':kt:godot-library:generateAPI' as an input of ':kt:godot-coroutine-library:sourcesJar'.
    2. Declare an explicit dependency on ':kt:godot-library:generateAPI' from ':kt:godot-coroutine-library:sourcesJar' using Task#dependsOn.
    3. Declare an explicit dependency on ':kt:godot-library:generateAPI' from ':kt:godot-coroutine-library:sourcesJar' using Task#mustRunAfter.

File: kt/api-generator/src/main/kotlin/godot/codegen/extensions/TypedExtensions.kt (hunk 4 above)
Comment (CedNaru) [reply kt/api-generator/src/main/kotlin/godot/codegen/extensions/TypedExtensions.kt:67]: Yes it was. Callable and Signal have been moved to the same package as the other coretypes. For consistency. They were initially in different packages back in Godot 3, but now they are full fledged Variant types so I moved them to their rightful place with the others.

**Review by piiertho — APPROVED** [review]

**Review by chippmann — APPROVED** [review]
What i meant with tests which switch coroutine context is that you await a signal in a test using `withContext(Dispatcher.SomeOtherdispatcherThanYouStartedWith) { signal.await() }` But that can be added later as well.

Other than that I'm happy :-) Good first implementation with which we can start with and have a look how it feels.
